### PR TITLE
Dpr2 666 execute query api

### PR DESCRIPTION
**Pull request by gavriil-g-moj**
The following PR is creating the asynchronous execute statement API using Redshift data.
Note: This is pending cross account connectivity so the call will not work until this is resolved.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportinglib/controller/ConfiguredApiController.kt [hunk 1]
```diff
@@ -113,6 +113,66 @@ class ConfiguredApiController(val configuredApiService: ConfiguredApiService) {
...
+            reportId = reportId,
+            reportVariantId = reportVariantId,
+            filters = filtersOnly(filters),
+            sortColumn = sortColumn,
+            sortedAsc = sortedAsc,
+            userToken = if (authentication is DprAuthAwareAuthenticationToken) authentication else null,
+            dataProductDefinitionsPath = dataProductDefinitionsPath,
+          ),
+        )
+    } catch (exception: NoDataAvailableException) {
```
Comment (gavriil-g-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportinglib/controller/ConfiguredApiController.kt:165]: We might need to think if this is how we want to handle all exceptions on the frontend for this API or have some specific cases.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportinglib/data/ConfiguredApiRepository.kt [hunk 2]
```diff
@@ -7,11 +7,18 @@ import org.springframework.context.ApplicationContext
 import org.springframework.jdbc.core.namedparam.MapSqlParameterSource
 import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
 import org.springframework.stereotype.Service
+import software.amazon.awssdk.services.redshiftdata.RedshiftDataClient
+import software.amazon.awssdk.services.redshiftdata.model.ExecuteStatementRequest
+import software.amazon.awssdk.services.redshiftdata.model.ExecuteStatementResponse
+import software.amazon.awssdk.services.redshiftdata.model.SqlParameter
```
Comment (stalwartsolutions) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportinglib/data/ConfiguredApiRepository.kt:13]: Do you think it's worth separating this into two repositories? Possibly with a shared parent or helper class?

They're two quite different mechanisms, using two different sources.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportinglib/data/ConfiguredApiRepositoryTest.kt [hunk 3]
```diff
@@ -2,15 +2,22 @@ package uk.gov.justice.digital.hmpps.digitalprisonreportinglib.data
```
Comment (stalwartsolutions) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportinglib/data/ConfiguredApiRepositoryTest.kt:2]: Is it feasible to add an integration test for this too? To ensure the full path from the API to the repo is happy?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportinglib/data/ConfiguredApiRepository.kt (hunk 2 above)
Comment (gavriil-g-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportinglib/data/ConfiguredApiRepository.kt:13]: Indeed I had thought of that and decided to put them in the same repository initially and confirm we can call this successfully first and then separate them.
But I will look into this again to make the separation from the beginning.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportinglib/data/ConfiguredApiRepositoryTest.kt (hunk 3 above)
Comment (gavriil-g-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportinglib/data/ConfiguredApiRepositoryTest.kt:2]: I have tried this unsuccessfully as the only container I have found to use for testing is LocalStack.
I used [this with testcontainers](https://java.testcontainers.org/modules/localstack/) and there were two issues.
First of all it seemed unreliable as when I replaced Postgress with LocalStack the test we have in ConfiguredApiRepositoryTestContainersTest to catch a Redshift specific error would pass if I changed the code back to what was causing that bug. This test failed with the Postgres container as it should.
And secondly, which is the main blocker for that, running this test locally would not work as authentication fails when trying to connect to that LocalStack container.
I have been trying to test the Controller with a mocked service to confirm the controller calls the service but there is an issue with @MockBean not replacing the service with the mocked one in the Spring @WebMvcTest.
I will spend some more time on this and see if I can find an alternative.

Comment (stalwartsolutions) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportinglib/data/ConfiguredApiRepositoryTest.kt:2]: OK, thanks. If it's going to be a pig, then don't worry for this PR.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportinglib/data/ConfiguredApiRepository.kt (hunk 2 above)
Comment (gavriil-g-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportinglib/data/ConfiguredApiRepository.kt]: The repositories have been separated and inheriting from a common helper class.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportinglib/data/ConfiguredApiRepositoryTest.kt (hunk 3 above)
Comment (gavriil-g-moj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/digitalprisonreportinglib/data/ConfiguredApiRepositoryTest.kt]: I spent some more time but this test still has issues. 
We could merge this PR and look into this further as part of future changes.

**Review by stalwartsolutions — APPROVED** [review]

**Pull request merged**